### PR TITLE
Added detect hidden inode feature

### PR DESCRIPTION
**Pull request by 0xF41**
- added some KERN_INFO messages in the main LKM
- added hidden-inode-detector.py and detectinodes.c
- filled up switch statement in user app 
- add some stuff in .gitignore

gcc client.c $(python3.5-config --cflags)  $(python3.5-config --ldflags) -o client

extra args in gcc statement needed for C to interface with python

### CODE REVIEW

File: client.c [hunk 1]
```diff
@@ -67,19 +73,46 @@ int main(int argc, char **argv)
...
          }
          checkallquick();
+
+         // Log to LKM
+         memset(cmd_buf, 0x0, BUF_SIZE);
+         sprintf(cmd_buf, DETECTPID_CMD);
+         if (write(fd, cmd_buf, strlen(cmd_buf)) < 0){
+            __err("[__ERROR_2__]", perror, -1);
+         }
+         else {
```
Comment (lckjosh) [inline client.c]: remove else for safety

File: client.c [hunk 2]
```diff
@@ -67,19 +73,46 @@ int main(int argc, char **argv)
...
+            break;
+         }
+         else {
+            // Log to LKM
+            memset(cmd_buf, 0x0, BUF_SIZE);
+            sprintf(cmd_buf, DETECTINODE_CMD);
+            if (write(fd, cmd_buf, strlen(cmd_buf)) < 0){
+               __err("[__ERROR_2__]", perror, -1);
+            }
+            else {
```
Comment (lckjosh) [inline client.c]: remove else for safety

File: detectiontool.c [hunk 3]
```diff
@@ -74,23 +74,29 @@ static ssize_t tool_procfs_write(struct file *fp,
 
 	if (strcmp(buf, DETECTPID_CMD) == 0)
 	{
-		// detect hidden pids
+		// detect hidden pids (Userspace)
+		printk(KERN_INFO "detection tool: Detect hidden pids (-p) has been invoked by the user.\n");
+		memset(buf, 0x0, strlen(buf));
```
Comment (lckjosh) [inline detectiontool.c]: clear whole array [16] just to be safe

File: detectiontool.c [hunk 4]
```diff
@@ -74,23 +74,29 @@ static ssize_t tool_procfs_write(struct file *fp,
...
 	}
 
-	else if (strcmp(buf, DETECTFILE_CMD) == 0)
+	else if (strcmp(buf, DETECTINODE_CMD) == 0)
 	{
-		// detect hidden files
+		// detect hidden files (Userspace)
+		printk(KERN_INFO "detection tool: Detect hidden inode scan (-f) has been invoked by the user.\n");
+		memset(buf, 0x0, strlen(buf));
```
Comment (lckjosh) [inline detectiontool.c]: clear whole array [16] just to be safe

File: client.c [hunk 5]
```diff
@@ -67,19 +73,46 @@ int main(int argc, char **argv)
...
       case 'p':
+         // detect hidden processes
          if (getuid() != 0)
          {
             printf("You must be root to perform this function!\n");
             exit(1);
          }
          checkallquick();
+
+         // Log to LKM
```
Comment (lckjosh) [inline client.c:85]: log first, then run `checkallquick()`, same as -s and -m

File: client.c [hunk 6]
```diff
@@ -67,19 +73,46 @@ int main(int argc, char **argv)
...
          break;
       case 'f':
-         // detect hidden files
+         // detect hidden inodes
+         if (hideinodedetector() != 0){
+            printf("client.c: Python script failed to execute completely.\n");
+            break;
+         }
+         else {
+            // Log to LKM
```
Comment (lckjosh) [inline client.c]: log first, then run `hideinodedetector()`

**Review by lckjosh — CHANGES_REQUESTED** [review]
```
Traceback (most recent call last):
  File "hidden-inode-detector.py", line 260, in <module>
    process = subprocess.Popen(['ffind', '-u', str(sys.argv[1]), str(anomaly)], stdout=subprocess.PIPE, stderr=subprocess.PIPE)
  File "/usr/lib/python3.5/subprocess.py", line 947, in __init__
    restore_signals, start_new_session)
  File "/usr/lib/python3.5/subprocess.py", line 1551, in _execute_child
    raise child_exception_type(errno_num, err_msg)
FileNotFoundError: [Errno 2] No such file or directory: 'ffind'
filedirdetector.c: Python script hidden-inode-detector.py raised an exception. Hidden inode scan is incomplete.
client.c: Python script failed to execute completely.
```

error message when testing for lilyofthevalley, ubuntu 16.04

pls include all installation instructions in README e.g. installation of python3, pip3, etc

File: client.c [hunk 7]
```diff
@@ -61,6 +61,9 @@ A rootkit may still be present on your system but is not hiding any process at t
 
 int main(int argc, char **argv)
 {
+   if (argc < 2)
+      usage_err();
```
Comment (0xF41) [inline client.c:65]: Note: Quick check if user supplied an option otherwise print usage_err()

File: client.c (hunk 5 above)
Comment (0xF41) [reply client.c:85]: Placed after the write() part

**Comment by 0xF41** [comment]
Modifications to "changed stuff before adding detect hidden inode feature" commit:
- Added quick check to client.c if user has supplied an option 
- Updated README.md dependency section + gcc command
- Removed else block in user-space switch statement
- Rearranged so that user-space functions are called after logging to LKM 
- Changed a bit of print() lines in the python script

**Pull request merged**